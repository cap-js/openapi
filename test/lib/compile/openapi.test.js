const toOpenApi = require('../../../lib/compile');
const cds = require('@sap/cds')

const someOpenApi = { openapi: '3.0.2', info: {}, servers: [{}], tags: [{}], paths: {}, components: {} }
const SCENARIO = Object.freeze({
  positive: 'positive',
  notAllowedAnnotations: 'notAllowed',
  notMatchingValues: 'notMatching',
  checkProperty: 'checkProperty'
})

function checkAnnotations(csn, annotations, scenario = SCENARIO.positive, property = '') {
  const openApi = toOpenApi(csn)
  const schemas = Object.entries(openApi.components.schemas).filter(([key]) => key.startsWith('sap.odm.test.A.E1'))

  // Test if the openAPI document was generated with some schemas.
  expect(openApi).toBeDefined()
  expect(openApi.components.schemas).toBeDefined()
  expect(schemas.length > 0).toBeTruthy()

  // Expect that not-allowed ODM annotations are unavailable in the schema.
  if (scenario === SCENARIO.notAllowedAnnotations) {
    for (const [, schema] of schemas) {
      for (const [annKey] of annotations) {
        expect(schema[annKey]).not.toBeDefined()
      }
    }
    return;
  }

  // Expect that even the ODM annotations with not-matched values will be derived.
  if (scenario === SCENARIO.notMatchingValues) {
    for (const [, schema] of schemas) {
      for (const [annKey, annValue] of annotations) {
        expect(schema[annKey]).toBe(annValue)
      }
    }
    return;
  }

  if (scenario === SCENARIO.checkProperty) {
    for (const [, schema] of schemas) {
      const propertyObj = schema.properties[property]
      for (const [annKey, annValue] of annotations) {
        expect(propertyObj[annKey]).toBe(annValue)
      }
    }
    return
  }

  for (const [, schema] of schemas) {
    for (const [annKey, annValue] of annotations) {
      expect(schema[annKey]).toBe(annValue)
    }
  }

  // Test that no other places contain the ODM extensions in the OpenAPI document.

  // components.schemas where the schemas are not from entity E1.
  const notE1 = Object.entries(openApi.components.schemas).filter(([key]) => !key.startsWith('sap.odm.test.A.E1'))
  for (const [, schema] of notE1) {
    const schemaString = JSON.stringify(schema)
    for (const [annKey] of annotations) {
      expect(schemaString).not.toContain(annKey)
    }
  }

  // all other components of the OpenAPI document except the schemas.
  const openApiNoSchemas = JSON.stringify({ ...openApi, components: { parameters: { ...openApi.components.parameters }, responses: { ...openApi.components.responses } } })
  for (const [annKey] of annotations) {
    expect(openApiNoSchemas).not.toContain(annKey)
  }
}

describe('OpenAPI export', () => {
  let _precompile_edms
  beforeAll(() => {
    _precompile_edms = cds.env.features.precompile_edms
    cds.env.features.precompile_edms = false
  })
  afterAll(() => {
    cds.env.features.precompile_edms = _precompile_edms
  })

  test('one service', () => {
    const csn = cds.compile.to.csn(`
      service A {entity E { key ID : UUID; };};`
    );
    const openapi = toOpenApi(csn);
    expect(openapi).toMatchObject(someOpenApi);
    // UUID elements are not required
    expect(openapi.components.schemas['A.E-create']).not.toHaveProperty('required');
  });

  test('one service, namespace', () => {
    const csn = cds.compile.to.csn(`
      namespace com.sap;
      service A {entity E { key ID : UUID; };};`
    );
    const openapi = toOpenApi(csn);
    expect(openapi).toMatchObject(someOpenApi);
  });

  test('one service, multiple protocols', () => {
    const csn = cds.compile.to.csn(`
      namespace com.sap;
      @protocol: ['odata', 'rest']
      service A {entity E { key ID : UUID; };};`
    );
    const openapi = toOpenApi(csn);
    const filesFound = new Set();
    for (const [content, metadata] of openapi) {
      expect(content).toMatchObject(someOpenApi);
      filesFound.add(metadata.file);
    }
    expect(filesFound).toMatchObject(new Set(['.odata', '.rest']));
  });

  test('multiple services', () => {
    const csn = cds.compile.to.csn(`
      service A {entity E { key ID : UUID; };};
      service B {entity F { key ID : UUID; };};`
    )
    expect(() => toOpenApi(csn, { service: 'foo' })).toThrowError(/no service/si)

    let openapi = toOpenApi(csn, { service: 'A' });
    expect(openapi).toMatchObject(someOpenApi);

    openapi = toOpenApi(csn, { service: 'B' });
    expect(openapi).toMatchObject(someOpenApi);

    openapi = toOpenApi(csn, { service: 'all' });
    const filesFound = new Set();
    for (const [content, metadata] of openapi) {
      expect(content).toMatchObject(someOpenApi);
      filesFound.add(metadata.file);
    }
    expect(filesFound).toMatchObject(new Set(['A', 'B']));
  });

  test('multiple services, namespace', () => {
    const csn = cds.compile.to.csn(`
      namespace com.sap;
      service A {entity E { key ID : UUID; };};
      service B {entity F { key ID : UUID; };};`
    )
    expect(() => toOpenApi(csn, { service: 'foo' })).toThrowError(/no service/si)

    let openapi = toOpenApi(csn, { service: 'com.sap.A' });
    expect(openapi).toMatchObject(someOpenApi);

    openapi = toOpenApi(csn, { service: 'com.sap.B' });
    expect(openapi).toMatchObject(someOpenApi);

    openapi = toOpenApi(csn, { service: 'all' });
    const filesFound = new Set();
    for (const [content, metadata] of openapi) {
      expect(content).toMatchObject(someOpenApi);
      filesFound.add(metadata.file);
    }
    expect(filesFound).toMatchObject(new Set(['com.sap.A', 'com.sap.B']));
  });

  test('multiple services, multiple protocols', () => {
    const csn = cds.compile.to.csn(`
      namespace com.sap;
      @protocol: ['odata', 'rest']
      service A {entity E { key ID : UUID; };};
      @protocol: ['odata', 'rest']
      service B {entity F { key ID : UUID; };};`
    );
    const openapi = toOpenApi(csn, { service: 'all' });
    const filesFound = new Set();
    for (const [content, metadata] of openapi) {
      expect(content).toMatchObject(someOpenApi);
      filesFound.add(metadata.file);
    }
    expect(filesFound).toMatchObject(new Set(['com.sap.A.odata', 'com.sap.A.rest', 'com.sap.B.odata', 'com.sap.B.rest']));
  });

  test('options: url', () => {
    const csn = cds.compile.to.csn(`
      @path:'/a' service A {entity E { key ID : UUID; };};
      service B {entity F { key ID : UUID; };};`
    );
    let openapi = toOpenApi(csn, { service: 'A' });
    expect(openapi).toMatchObject({ servers: [{ url: '/a' }] });
    expect(openapi.info.description).toBe("Use @Core.LongDescription: '...' on your CDS service to provide a meaningful description.")

    openapi = toOpenApi(csn, { service: 'A', 'openapi:url': 'http://foo.bar:8080' });
    expect(openapi).toMatchObject({ servers: [{ url: 'http://foo.bar:8080' }] });

    openapi = toOpenApi(csn, { service: 'A', 'openapi:url': 'http://foo.bar:8080//${service-path}/foo' });
    expect(openapi).toMatchObject({ servers: [{ url: 'http://foo.bar:8080/a/foo' }] });
  });

  test('options: diagram', () => {
    const csn = cds.compile.to.csn(`
      service A {entity E { key ID : UUID; };};`
    );
    let openapi = toOpenApi(csn);
    expect(openapi.info.description).not.toMatch(/yuml.*diagram/i)
    openapi = toOpenApi(csn, { 'openapi:diagram': true });
    expect(openapi.info.description).toMatch(/yuml.*diagram/i)
  });

  test('options: servers', () => {
    const csn = cds.compile.to.csn(`
      service A {entity E { key ID : UUID; };};`
    );
    const serverObj = "[{\n \"url\": \"https://{customerId}.saas-app.com:{port}/v2\",\n \"variables\": {\n \"customerId\": \"demo\",\n \"description\": \"Customer ID assigned by the service provider\"\n }\n}]"
    const openapi = toOpenApi(csn, { 'openapi:servers': serverObj });
    expect(openapi.servers).toBeTruthy();
  });

  test('options: Multiple servers', () => {
    const csn = cds.compile.to.csn(`
      service A {entity E { key ID : UUID; };};`
    );
    const serverObj = "[{\n \"url\": \"https://{customer1Id}.saas-app.com:{port}/v2\",\n \"variables\": {\n \"customer1Id\": \"demo\",\n \"description\": \"Customer1 ID assigned by the service provider\"\n }\n}, {\n \"url\": \"https://{customer2Id}.saas-app.com:{port}/v2\",\n \"variables\": {\n \"customer2Id\": \"demo\",\n \"description\": \"Customer2 ID assigned by the service provider\"\n }\n}]"
    const openapi = toOpenApi(csn, { 'openapi:servers': serverObj });
    expect(openapi.servers).toBeTruthy();
    expect(openapi.servers[0].url).toMatch('https://{customer1Id}.saas-app.com:{port}/v2/odata/v4/A')
  });


  test('options: servers - wrong JSON', () => {
    const csn = cds.compile.to.csn(`
      service A {entity E { key ID : UUID; };};`
    );
    const serverObj = "[{\n \"url\": \"https://{customerId}.saas-app.com:{port}/v2\",\n \"variables\":\": \"Customer ID assigned by the service provider\"\n }\n}]"
    try {
      toOpenApi(csn, { 'openapi:servers': serverObj });
    }
    catch (e) {
      expect(e.message).toBe("The input server object is invalid.");
    }
  });

  test('annotations: root entity property', () => {
    const csn = cds.compile.to.csn(`
      namespace sap.odm.test;
      service A {
        @ODM.root: true
        entity E1 { key id: String(4); }
        entity E2 { key id: String(4); }
      }
    `)

    const openAPI = toOpenApi(csn)
    expect(openAPI).toBeDefined()
    expect(openAPI.components.schemas["sap.odm.test.A.E1"]).toMatchObject({ "x-sap-root-entity": true })
    expect(openAPI.components.schemas["sap.odm.test.A.E1-create"]["x-sap-root-entity"]).toBeUndefined()
    expect(openAPI.components.schemas["sap.odm.test.A.E1-update"]["x-sap-root-entity"]).toBeUndefined()
  })

  test('odm annotations: entity name and oid property', () => {
    const csn = cds.compile.to.csn(`
      namespace sap.odm.test;
      service A {
        @ODM.entityName: 'sap.odm.test.E1'
        @ODM.oid: 'oid'
        entity E1 { key id: String(4); oid: String(128);}
        entity E2 { key id: String(4); }
      }
    `)
    checkAnnotations(
      csn,
      new Map([["x-sap-odm-entity-name", "sap.odm.test.E1"], ["x-sap-odm-oid", "oid"]]))
  })

  test('odm annotations: not valid names', () => {
    const csn = cds.compile.to.csn(`
      namespace sap.odm.test;
      service A {
        @ODM.entityNameInvalid: 'sap.odm.test.E1'
        @ODM.oidInvalid: 'oid'
        entity E1 { key id: String(4); oid: String(128);}
        entity E2 { key id: String(4); }
      }
    `)

    checkAnnotations(
      csn,
      new Map([["x-sap-odm-entity-name-not-allowed", "sap.odm.test.E1"], ["x-sap-odm-oid-not-allowed", "oid"]]),
      SCENARIO.notAllowedAnnotations
    )
  })

  test('odm annotations: @ODM.oid value has no matching property', () => {
    const csn = cds.compile.to.csn(`
      namespace sap.odm.test;
      service A {
        @ODM.oid: 'foo'
        entity E1 { key id: String(4); oid: String(128);}
        entity E2 { key id: String(4); }
      }
    `)

    checkAnnotations(
      csn,
      new Map([["x-sap-odm-oid", "foo"]]),
      SCENARIO.notMatchingValues
    )
  })

  test('odm annotations: @ODM.entityName value not matching the entity name', () => {
    const csn = cds.compile.to.csn(`
      namespace sap.odm.test;
      service A {
        @ODM.entityName: 'sap.odm.test.bar'
        entity E1 { key id: String(4); oid: String(128);}
        entity E2 { key id: String(4); }
      }
    `)

    checkAnnotations(
      csn,
      new Map([["x-sap-odm-entity-name", "sap.odm.test.bar"]]),
      SCENARIO.notMatchingValues
    )
  })

  test('odm annotations: @ODM.oidReference.entityName annotation is added to the schema', () => {
    const csn = cds.compile.to.csn(`
      namespace sap.odm.test;
      service A {
        entity E1 { 
          key id: String(4); 
          oid: String(128); 
          @ODM.oidReference.entityName: 'ReferencedEntityName'
          ref: Association to one E2;
        }
        entity E2 { key id: String(4); }
      }
    `)

    checkAnnotations(
      csn,
      new Map([["x-sap-odm-oid-reference-entity-name", "ReferencedEntityName"]]),
      SCENARIO.checkProperty,
      'ref_id'
    )
  })

  test('OpenAPI annotations: @OpenAPI.externalDocs annotation is added to the schema', () => {
    const csn = cds.compile.to.csn(`
      namespace sap.OpenAPI.test;
      @OpenAPI.externalDocs: {
        description: 'API Guide',
        url: 'https://help.sap.com/docs/product/123.html'
      }
      service A {
        entity E1 { 
          key id: String(4); 
          oid: String(128); 
        }
          }`);
        const openAPI = toOpenApi(csn);
        expect(openAPI.externalDocs).toBeDefined();
        expect(openAPI.externalDocs.description).toBe('API Guide');
        expect(openAPI.externalDocs.url).toBe('https://help.sap.com/docs/product/123.html');
  }
  );
});
