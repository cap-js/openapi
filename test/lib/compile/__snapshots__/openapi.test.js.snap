// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAPI export ER annotations er annotations is correct 1`] = `
{
  "components": {
    "parameters": {
      "count": {
        "description": "Include count of items, see [Count](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptioncount)",
        "in": "query",
        "name": "$count",
        "schema": {
          "type": "boolean",
        },
      },
      "search": {
        "description": "Search items by search phrases, see [Searching](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptionsearch)",
        "in": "query",
        "name": "$search",
        "schema": {
          "type": "string",
        },
      },
      "skip": {
        "description": "Skip the first n items, see [Paging - Skip](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptionskip)",
        "in": "query",
        "name": "$skip",
        "schema": {
          "minimum": 0,
          "type": "integer",
        },
      },
      "top": {
        "description": "Show only the first n items, see [Paging - Top](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptiontop)",
        "example": 50,
        "in": "query",
        "name": "$top",
        "schema": {
          "minimum": 0,
          "type": "integer",
        },
      },
    },
    "responses": {
      "error": {
        "content": {
          "application/json": {
            "schema": {
              "$ref": "#/components/schemas/error",
            },
          },
        },
        "description": "Error",
      },
    },
    "schemas": {
      "A.Material": {
        "properties": {
          "BOM": {
            "maxLength": 30,
            "nullable": true,
            "type": "string",
            "x-entity-relationship-reference": {
              "referencedEntityType": "sap.vdm.sont:BillOfMaterial",
              "referencedPropertyType": "sap.vdm.gfn:BillOfMaterialId",
            },
          },
          "ObjectID": {
            "maxLength": 18,
            "type": "string",
            "x-entity-relationship-property-type": "sap.vdm.gfn:MaterialId",
          },
          "PurchaseOrder": {
            "example": "01234567-89ab-cdef-0123-456789abcdef",
            "format": "uuid",
            "nullable": true,
            "type": "string",
            "x-entity-relationship-reference": {
              "referencedEntityType": "sap.sm:PurchaseOrder",
              "referencedPropertyType": "sap.sm:PurchaseOrderUUID",
            },
            "x-sap-odm-oid-reference-entity-name": "PurchaseOrder",
          },
          "manufacturer": {
            "maxLength": 40,
            "nullable": true,
            "type": "string",
            "x-entity-relationship-reference": {
              "referencedEntityType": "sap.vdm.sont:BusinessPartner",
              "referencedPropertyType": "sap.vdm.gfn::BusinessPartnerNumber",
            },
          },
        },
        "title": "Material",
        "type": "object",
        "x-entity-relationship-entity-ids": [
          {
            "propertyTypes": [
              "sap.vdm.gfn:MaterialId",
            ],
          },
        ],
        "x-entity-relationship-entity-type": "sap.vdm.sont:Material",
        "x-sap-odm-entity-name": "Product",
        "x-sap-odm-oid": "oid",
      },
      "A.Material-create": {
        "properties": {
          "BOM": {
            "maxLength": 30,
            "nullable": true,
            "type": "string",
            "x-entity-relationship-reference": {
              "referencedEntityType": "sap.vdm.sont:BillOfMaterial",
              "referencedPropertyType": "sap.vdm.gfn:BillOfMaterialId",
            },
          },
          "ObjectID": {
            "maxLength": 18,
            "type": "string",
            "x-entity-relationship-property-type": "sap.vdm.gfn:MaterialId",
          },
          "PurchaseOrder": {
            "example": "01234567-89ab-cdef-0123-456789abcdef",
            "format": "uuid",
            "nullable": true,
            "type": "string",
            "x-entity-relationship-reference": {
              "referencedEntityType": "sap.sm:PurchaseOrder",
              "referencedPropertyType": "sap.sm:PurchaseOrderUUID",
            },
            "x-sap-odm-oid-reference-entity-name": "PurchaseOrder",
          },
          "manufacturer": {
            "maxLength": 40,
            "nullable": true,
            "type": "string",
            "x-entity-relationship-reference": {
              "referencedEntityType": "sap.vdm.sont:BusinessPartner",
              "referencedPropertyType": "sap.vdm.gfn::BusinessPartnerNumber",
            },
          },
        },
        "required": [
          "ObjectID",
        ],
        "title": "Material (for create)",
        "type": "object",
        "x-entity-relationship-entity-ids": [
          {
            "propertyTypes": [
              "sap.vdm.gfn:MaterialId",
            ],
          },
        ],
        "x-entity-relationship-entity-type": "sap.vdm.sont:Material",
        "x-sap-odm-entity-name": "Product",
        "x-sap-odm-oid": "oid",
      },
      "A.Material-update": {
        "properties": {
          "BOM": {
            "maxLength": 30,
            "nullable": true,
            "type": "string",
            "x-entity-relationship-reference": {
              "referencedEntityType": "sap.vdm.sont:BillOfMaterial",
              "referencedPropertyType": "sap.vdm.gfn:BillOfMaterialId",
            },
          },
          "PurchaseOrder": {
            "example": "01234567-89ab-cdef-0123-456789abcdef",
            "format": "uuid",
            "nullable": true,
            "type": "string",
            "x-entity-relationship-reference": {
              "referencedEntityType": "sap.sm:PurchaseOrder",
              "referencedPropertyType": "sap.sm:PurchaseOrderUUID",
            },
            "x-sap-odm-oid-reference-entity-name": "PurchaseOrder",
          },
          "manufacturer": {
            "maxLength": 40,
            "nullable": true,
            "type": "string",
            "x-entity-relationship-reference": {
              "referencedEntityType": "sap.vdm.sont:BusinessPartner",
              "referencedPropertyType": "sap.vdm.gfn::BusinessPartnerNumber",
            },
          },
        },
        "title": "Material (for update)",
        "type": "object",
        "x-entity-relationship-entity-ids": [
          {
            "propertyTypes": [
              "sap.vdm.gfn:MaterialId",
            ],
          },
        ],
        "x-entity-relationship-entity-type": "sap.vdm.sont:Material",
        "x-sap-odm-entity-name": "Product",
        "x-sap-odm-oid": "oid",
      },
      "count": {
        "anyOf": [
          {
            "type": "number",
          },
          {
            "type": "string",
          },
        ],
        "description": "The number of entities in the collection. Available when using the [$count](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptioncount) query option.",
      },
      "error": {
        "properties": {
          "error": {
            "properties": {
              "code": {
                "type": "string",
              },
              "details": {
                "items": {
                  "properties": {
                    "code": {
                      "type": "string",
                    },
                    "message": {
                      "type": "string",
                    },
                    "target": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "code",
                    "message",
                  ],
                  "type": "object",
                },
                "type": "array",
              },
              "innererror": {
                "description": "The structure of this object is service-specific",
                "type": "object",
              },
              "message": {
                "type": "string",
              },
              "target": {
                "type": "string",
              },
            },
            "required": [
              "code",
              "message",
            ],
            "type": "object",
          },
        },
        "required": [
          "error",
        ],
        "type": "object",
      },
    },
  },
  "info": {
    "description": "Use @Core.LongDescription: '...' on your CDS service to provide a meaningful description.",
    "title": "Use @title: '...' on your CDS service to provide a meaningful title.",
    "version": "",
  },
  "openapi": "3.0.2",
  "paths": {
    "/$batch": {
      "post": {
        "description": "Group multiple requests into a single request payload, see [Batch Requests](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_BatchRequests).

*Please note that "Try it out" is not supported for this request.*",
        "requestBody": {
          "content": {
            "multipart/mixed;boundary=request-separator": {
              "example": "--request-separator
Content-Type: application/http
Content-Transfer-Encoding: binary

GET Material HTTP/1.1
Accept: application/json


--request-separator--",
              "schema": {
                "type": "string",
              },
            },
          },
          "description": "Batch request",
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "multipart/mixed": {
                "example": "--response-separator
Content-Type: application/http

HTTP/1.1 200 OK
Content-Type: application/json

{...}
--response-separator--",
                "schema": {
                  "type": "string",
                },
              },
            },
            "description": "Batch response",
          },
          "4XX": {
            "$ref": "#/components/responses/error",
          },
        },
        "summary": "Sends a group of requests",
        "tags": [
          "Batch Requests",
        ],
      },
    },
    "/Material": {
      "get": {
        "parameters": [
          {
            "$ref": "#/components/parameters/top",
          },
          {
            "$ref": "#/components/parameters/skip",
          },
          {
            "$ref": "#/components/parameters/search",
          },
          {
            "description": "Filter items by property values, see [Filtering](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptionfilter)",
            "in": "query",
            "name": "$filter",
            "schema": {
              "type": "string",
            },
          },
          {
            "$ref": "#/components/parameters/count",
          },
          {
            "description": "Order items by property values, see [Sorting](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptionorderby)",
            "explode": false,
            "in": "query",
            "name": "$orderby",
            "schema": {
              "items": {
                "enum": [
                  "ObjectID",
                  "ObjectID desc",
                  "manufacturer",
                  "manufacturer desc",
                  "PurchaseOrder",
                  "PurchaseOrder desc",
                  "BOM",
                  "BOM desc",
                ],
                "type": "string",
              },
              "type": "array",
              "uniqueItems": true,
            },
          },
          {
            "description": "Select properties to be returned, see [Select](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptionselect)",
            "explode": false,
            "in": "query",
            "name": "$select",
            "schema": {
              "items": {
                "enum": [
                  "ObjectID",
                  "manufacturer",
                  "PurchaseOrder",
                  "BOM",
                ],
                "type": "string",
              },
              "type": "array",
              "uniqueItems": true,
            },
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "@count": {
                      "$ref": "#/components/schemas/count",
                    },
                    "value": {
                      "items": {
                        "$ref": "#/components/schemas/A.Material",
                      },
                      "type": "array",
                    },
                  },
                  "title": "Collection of Material",
                  "type": "object",
                },
              },
            },
            "description": "Retrieved material",
          },
          "4XX": {
            "$ref": "#/components/responses/error",
          },
        },
        "summary": "Retrieves a list of material.",
        "tags": [
          "Material",
        ],
      },
      "post": {
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/A.Material-create",
              },
            },
          },
          "description": "New material",
          "required": true,
        },
        "responses": {
          "201": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/A.Material",
                },
              },
            },
            "description": "Created material",
          },
          "4XX": {
            "$ref": "#/components/responses/error",
          },
        },
        "summary": "Creates a single material.",
        "tags": [
          "Material",
        ],
      },
    },
    "/Material('{ObjectID}')": {
      "delete": {
        "responses": {
          "204": {
            "description": "Success",
          },
          "4XX": {
            "$ref": "#/components/responses/error",
          },
        },
        "summary": "Deletes a single material.",
        "tags": [
          "Material",
        ],
      },
      "get": {
        "parameters": [
          {
            "description": "Select properties to be returned, see [Select](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptionselect)",
            "explode": false,
            "in": "query",
            "name": "$select",
            "schema": {
              "items": {
                "enum": [
                  "ObjectID",
                  "manufacturer",
                  "PurchaseOrder",
                  "BOM",
                ],
                "type": "string",
              },
              "type": "array",
              "uniqueItems": true,
            },
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/A.Material",
                },
              },
            },
            "description": "Retrieved material",
          },
          "4XX": {
            "$ref": "#/components/responses/error",
          },
        },
        "summary": "Retrieves a single material.",
        "tags": [
          "Material",
        ],
      },
      "parameters": [
        {
          "description": "key: ObjectID",
          "in": "path",
          "name": "ObjectID",
          "required": true,
          "schema": {
            "maxLength": 18,
            "type": "string",
            "x-entity-relationship-property-type": "sap.vdm.gfn:MaterialId",
          },
        },
      ],
      "patch": {
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/A.Material-update",
              },
            },
          },
          "description": "New property values",
          "required": true,
        },
        "responses": {
          "204": {
            "description": "Success",
          },
          "4XX": {
            "$ref": "#/components/responses/error",
          },
        },
        "summary": "Changes a single material.",
        "tags": [
          "Material",
        ],
      },
    },
  },
  "servers": [
    {
      "url": "/rest/a",
    },
  ],
  "tags": [
    {
      "name": "Material",
    },
  ],
  "x-odata-version": "4.01",
  "x-sap-api-type": "ODATAV4",
  "x-sap-shortText": "Use @Core.Description: '...' on your CDS service to provide a meaningful short text.",
}
`;
